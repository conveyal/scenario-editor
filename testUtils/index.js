export function mockComponents (components) {
  if (components instanceof Array) {
    const mockedComponents = {}
    for (let i = 0; i < components.length; i++) {
      const name = components[i]
      mockedComponents[name] = name
    }
    return mockedComponents
  }
}
