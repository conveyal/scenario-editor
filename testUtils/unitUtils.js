/* globals describe, expect, it */

import React from 'react'
import renderer from 'react-test-renderer'

/**
 * Performs a basic render test of a Component
 *
 * @param  {Mixed} components Can be either an array of objects or a single object
 *   Each object can have the following properties:
 *   - children: Any children to render within the component
 *   - component (required): The component to render
 *   - name: How to classify the test in the `describe section`
 *   - notToBeCalledFns: An array of String keys that map to property values that are jest.fn()'s that will be asserted as not called during basic render test
 *   - props (required): The props of the component to render
 *   - toBeCalledFns: Same as `notToBeCalledFns` except they are asserted to be called
 */
export function basicRenderTest (components) {
  // prepare components to be tested
  if (!Array.isArray(components)) {
    components = [components]
  }

  // run tests on each component
  components.forEach((component) => {
    const Component = component.component

    if (typeof component.name !== 'string') {
      throw new Error('Component name not provided in basicRenderTest cfg')
    }

    describe(`Component > ${component.name}`, () => {
      it('renders correctly', () => {
        const tree = renderer.create(
          <Component
            {...component.props}
            >
            {component.children}
          </Component>
        ).toJSON()
        expect(tree).toMatchSnapshot()

        if (Array.isArray(component.toBeCalledFns)) {
          component.toBeCalledFns.forEach((fn) => {
            expect(component.props[fn]).toBeCalled()
          })
        }

        if (Array.isArray(component.notToBeCalledFns)) {
          component.notToBeCalledFns.forEach((fn) => {
            expect(component.props[fn]).not.toBeCalled()
          })
        }
      })
    })
  })
}

/**
 * Creates a object to mock a module with multiple exports
 *
 * @param  {Array} components An array of Strings representing the named exports
 * @return {Object}           A mock representation of the named exports
 */
export function mockExports (components) {
  if (Array.isArray(components)) {
    const mockedComponents = {}
    for (let i = 0; i < components.length; i++) {
      const name = components[i]
      mockedComponents[name] = name
    }
    return mockedComponents
  }
}
