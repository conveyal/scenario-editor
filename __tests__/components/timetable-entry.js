/* global describe, it, expect, jest */

import React from 'react'
import renderer from 'react-test-renderer'

import { mockComponents } from '../../testUtils'

jest.mock('../../lib/components/input', () => { return mockComponents(['Checkbox', 'Number', 'Text']) })

import TimetableEntry from '../../lib/components/timetable-entry'

describe('TimetableEntry', () => {
  it('renders correctly', () => {
    const timetable = {
      monday: true,
      tuesday: true,
      wednesday: true,
      thursday: true,
      friday: true,
      saturday: false,
      sunday: false,
      headwaySecs: 900, // 15 minutes
      startTime: 28800, // 8am
      endTime: 57600 // 4pm
    }
    const replaceTimetableFn = jest.fn()
    const tree = renderer.create(
      <TimetableEntry
        timetable={timetable}
        replaceTimetable={replaceTimetableFn}
        />
    ).toJSON()
    expect(tree).toMatchSnapshot()
    expect(replaceTimetableFn).not.toBeCalled
  })
})
