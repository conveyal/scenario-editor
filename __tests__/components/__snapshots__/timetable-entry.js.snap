exports[`TimetableEntry renders correctly 1`] = `
<div>
  <div
    className="form-inline">
    <Checkbox
      checked={true}
      label="Mon"
      onChange={[Function onChange]} />
    <Checkbox
      checked={true}
      label="Tue"
      onChange={[Function onChange]} />
    <Checkbox
      checked={true}
      label="Wed"
      onChange={[Function onChange]} />
    <Checkbox
      checked={true}
      label="Thu"
      onChange={[Function onChange]} />
    <Checkbox
      checked={true}
      label="Fri"
      onChange={[Function onChange]} />
    <Checkbox
      checked={false}
      label="Sat"
      onChange={[Function onChange]} />
    <Checkbox
      checked={false}
      label="Sun"
      onChange={[Function onChange]} />
  </div>
  <Number
    label="Frequency (minutes)"
    onChange={[Function anonymous]}
    value={15} />
  <Text
    label="Start time"
    onChange={[Function anonymous]}
    type="time"
    value="08:00" />
  <Text
    label="End time"
    onChange={[Function anonymous]}
    type="time"
    value="16:00" />
</div>
`;
